apps/prairielearn/src/pages/instructorCourseAdminStaff/instructorCourseAdminStaff.js

```diff
@@ -14,365 +14,343 @@ const debug = require('debug')('prairielearn:' + path.basename(__filename, '.js'
-    // because we want to discourage the assignment of this role to many users
-    if (!['None', 'Previewer', 'Viewer', 'Editor'].includes(req.body.course_role)) {
-      return next(error.make(400, `Invalid requested course role: ${req.body.course_role}`));
-    }
+      // Verify the requested course role is valid - we choose to disallow Owner
+      // because we want to discourage the assignment of this role to many users
+      if (!['None', 'Previewer', 'Viewer', 'Editor'].includes(req.body.course_role)) {
+        throw error.make(400, `Invalid requested course role: ${req.body.course_role}`);
+      }
 
-    // Verify the course instance id associated with the requested course instance
-    // role is valid (should such a role have been requested)
-    let course_instance = null;
-    if (req.body.course_instance_id) {
-      course_instance = res.locals.authz_data.course_instances.find((ci) =>
-        idsEqual(ci.id, req.body.course_instance_id),
-      );
-      if (!course_instance) return next(error.make(400, `Invalid requested course instance role`));
-    }
+      // Verify the course instance id associated with the requested course instance
+      // role is valid (should such a role have been requested)
+      let course_instance = null;
+      if (req.body.course_instance_id) {
+        course_instance = res.locals.authz_data.course_instances.find((ci) =>
+          idsEqual(ci.id, req.body.course_instance_id),
+        );
+        if (!course_instance) {
+          throw error.make(400, `Invalid requested course instance role`);
+        }
+      }
 
-    // Verify the requested course instance role is valid
-    if (
-      course_instance &&
-      !['Student Data Viewer', 'Student Data Editor'].includes(req.body.course_instance_role)
-    ) {
-      return next(
-        error.make(400, `Invalid requested course instance role: ${req.body.course_instance_role}`),
-      );
-    }
-    // Iterate through UIDs
-    async.reduce(
-      uids,
-      { given_cp: [], not_given_cp: [], not_given_cip: [], errors: [] },
-      (memo, uid, callback) => {
-        const c_params = [
-          res.locals.course.id,
-          uid,
-          req.body.course_role,
-          res.locals.authz_data.authn_user.user_id,
-        ];
-        sqldb.call('course_permissions_insert_by_user_uid', c_params, (err, result) => {
-          if (
-            ERR(err, (e) => logger.verbose(`Failed to insert course permission for uid: ${uid}`, e))
-          ) {
+      // Verify the requested course instance role is valid
+      if (
+        course_instance &&
+        !['Student Data Viewer', 'Student Data Editor'].includes(req.body.course_instance_role)
+      ) {
+        throw error.make(
+          400,
+          `Invalid requested course instance role: ${req.body.course_instance_role}`,
+        );
+      }
+      // Iterate through UIDs
+      const result = await async.reduce(
```

Comment:
This block was probably the trickiest refactor here, might be worth an extra close look during review.

Reply:
Looks good on a somewhat quick read through. I don't want you to change this but I think this is a bad use case for `reduce()` because it serializes a naturally parallel loop. Ideally we'd use `eachLimit()` and accumulate the `given_cp`, etc, results in variables from outside the loop.

Reply:
I think I agree with that. We can save a change for another PR, probably when this gets TS-ified.